File: CarePortal/CarePortal.Application/Services/MetadataService.cs
using CarePortal.Application.DTOs;
using CarePortal.Application.Interfaces;
using CarePortal.Domain.Enums;

namespace CarePortal.Application.Services;

public class MetadataService : IMetadataService
{
    private static readonly object _lock = new object();
    private static List<MetadataListDto>? _cachedMetadata;
    private static DateTime _lastCacheTime = DateTime.MinValue;
    private static readonly TimeSpan _cacheExpiration = TimeSpan.FromMinutes(5); // Cache for 5 minutes

    public List<MetadataListDto> GetMetadata()
    {
        lock (_lock)
        {
            // Check if cache is still valid
            if (_cachedMetadata != null && DateTime.UtcNow - _lastCacheTime < _cacheExpiration)
            {
                return _cachedMetadata;
            }

            // Generate new metadata
            var metadataList = new List<MetadataListDto>();
            
            // Add ActivityType enum
            metadataList.Add(CreateMetadataListDto("ACTIVITY_TYPE", typeof(ActivityType)));
            
            // Add IncidentStatus enum
            metadataList.Add(CreateMetadataListDto("INCIDENT_STATUS", typeof(IncidentStatus)));
            
            // Add IncidentSeverity enum
            metadataList.Add(CreateMetadataListDto("INCIDENT_SEVERITY", typeof(IncidentSeverity)));
            
            // Add UserRole enum
            metadataList.Add(CreateMetadataListDto("USER_ROLE", typeof(UserRole)));
            
            // Add DocumentType enum
            metadataList.Add(CreateMetadataListDto("DOCUMENT_TYPE", typeof(DocumentType)));
            
            // Add DocumentStatus enum
            metadataList.Add(CreateMetadataListDto("DOCUMENT_STATUS", typeof(DocumentStatus)));
            
            // Update cache
            _cachedMetadata = metadataList;
            _lastCacheTime = DateTime.UtcNow;
            
            return metadataList;
        }
    }
    
    private MetadataListDto CreateMetadataListDto(string type, Type enumType)
    {
        var metadata = Enum.GetValues(enumType)
            .Cast<Enum>()
            .Select(value => new MetadataDto
            {
                ParamKey = value.ToString(),
                ParamValue = GetDisplayName(value),
                ParamValueInt = Convert.ToInt32(value) // Add the numeric value
            })
            .ToList();
            
        return new MetadataListDto
        {
            Type = type,
            Metadata = metadata
        };
    }
    
    private string GetDisplayName(Enum value)
    {
        return value switch
        {
            ActivityType activityType => activityType.GetDisplayName(),
            IncidentStatus status => status.GetDisplayName(),
            IncidentSeverity severity => severity.GetDisplayName(),
            UserRole userRole => userRole.GetDisplayName(),
            DocumentType documentType => documentType.GetDisplayName(),
            DocumentStatus documentStatus => documentStatus.GetDisplayName(),
            _ => value.ToString()
        };
    }
} 

File: CarePortalUI/src/services/metadataService.ts
import { apiService } from './api';

export interface MetadataDto {
  paramKey: string;
  paramValue: string;
  paramValueInt: number; // Numeric enum value for automatic mapping
}

export interface MetadataListDto {
  type: string;
  metadata: MetadataDto[];
}

class MetadataService {
  private static cache: MetadataListDto[] | null = null;
  private static cacheTime: number = 0;
  private static readonly CACHE_DURATION = 5 * 60 * 1000; // 5 minutes in milliseconds

  private isCacheValid(): boolean {
    return MetadataService.cache !== null && (Date.now() - MetadataService.cacheTime) < MetadataService.CACHE_DURATION;
  }

  async getMetadata(): Promise<MetadataListDto[]> {
    if (this.isCacheValid()) {
      return MetadataService.cache!;
    }

    const metadata = await apiService.get<MetadataListDto[]>('/Metadata/get-metadata');
    
    // Update cache
    MetadataService.cache = metadata;
    MetadataService.cacheTime = Date.now();
    
    return metadata;
  }

  async getEnumData(enumType: string): Promise<MetadataDto[]> {
    const metadata = await this.getMetadata();
    const enumData = metadata.find(item => item.type === enumType);
    return enumData?.metadata || [];
  }

  async getActivityTypes(): Promise<MetadataDto[]> {
    return await this.getEnumData('ACTIVITY_TYPE');
  }

  async getIncidentStatuses(): Promise<MetadataDto[]> {
    return await this.getEnumData('INCIDENT_STATUS');
  }

  async getIncidentSeverities(): Promise<MetadataDto[]> {
    return await this.getEnumData('INCIDENT_SEVERITY');
  }

  async getUserRoles(): Promise<MetadataDto[]> {
    return await this.getEnumData('USER_ROLE');
  }

  async getDocumentTypes(): Promise<MetadataDto[]> {
    return await this.getEnumData('DOCUMENT_TYPE');
  }

  async getDocumentStatuses(): Promise<MetadataDto[]> {
    return await this.getEnumData('DOCUMENT_STATUS');
  }

  // Method to get enum value by key (replaces static enum mapping)
  async getEnumValueByKey(enumType: string, key: string): Promise<number> {
    const metadata = await this.getMetadata();
    const enumData = metadata.find(item => item.type === enumType);
    const item = enumData?.metadata.find(m => m.paramKey === key);
    return item?.paramValueInt || 0;
  }

  // Method to get enum key by value (reverse lookup)
  async getEnumKeyByValue(enumType: string, value: number): Promise<string> {
    const metadata = await this.getMetadata();
    const enumData = metadata.find(item => item.type === enumType);
    const item = enumData?.metadata.find(m => m.paramValueInt === value);
    return item?.paramKey || '';
  }

  // Method to clear cache if needed
  clearCache(): void {
    MetadataService.cache = null;
    MetadataService.cacheTime = 0;
  }
}

export const metadataService = new MetadataService(); 